src/NServiceBus.Core/Persistence/InMemory/SagaPersister/InMemorySagaPersister.cs

```diff
@@ -140,41 +129,60 @@ where s.Value.SagaData.GetType() == sagaType && (s.Key != saga.Id)
             }
         }
 
-        IContainSagaData DeepClone(IContainSagaData source)
-        {
-            var json = serializer.SerializeObject(source);
-            return (IContainSagaData) serializer.DeserializeObject(json, source.GetType());
-        }
-
         ConcurrentDictionary<Guid, VersionedSagaEntity> data = new ConcurrentDictionary<Guid, VersionedSagaEntity>();
-        JsonMessageSerializer serializer = new JsonMessageSerializer(null);
 
         int version;
 
         class VersionedSagaEntity
         {
-            public void RecordRead(IContainSagaData sagaEntity, int currentVersion)
+            public VersionedSagaEntity(IContainSagaData sagaData, VersionedSagaEntity original = null)
+            {
+                SagaData = DeepClone(sagaData);
+                VersionCache = original?.VersionCache ?? new ConditionalWeakTable<IContainSagaData, SagaVersion>();
+            }
+
+            public TSagaData Read<TSagaData>(long currentVersion)
+                where TSagaData : IContainSagaData
             {
-                VersionCache[sagaEntity] = currentVersion;
+                var clone = DeepClone(SagaData);
+                VersionCache.Add(clone, new SagaVersion(currentVersion));
+                return (TSagaData) clone;
             }
 
-            public void ConcurrencyCheck(IContainSagaData sagaEntity, int currentVersion)
+            public void ConcurrencyCheck(IContainSagaData sagaEntity, long currentVersion)
             {
-                int v;
+                SagaVersion v;
                 if (!VersionCache.TryGetValue(sagaEntity, out v))
                 {
                     throw new Exception($"InMemorySagaPersister in an inconsistent state: entity Id[{sagaEntity.Id}] not read.");
                 }
 
-                if (v != currentVersion)
+                if (v.version != currentVersion)
                 {
                     throw new Exception($@"InMemorySagaPersister concurrency violation: saga entity Id[{sagaEntity.Id}] already saved.");
                 }
             }
 
-            public IContainSagaData SagaData;
+            static IContainSagaData DeepClone(IContainSagaData source)
+            {
+                var json = serializer.SerializeObject(source);
+                return (IContainSagaData) serializer.DeserializeObject(json, source.GetType());
+            }
+
+            public readonly IContainSagaData SagaData;
```

Comment:
Not a huge fan of public fields, but since it's `readonly` I guess it's ok. If it's public, looks like we're going with TitleCase for the naming scheme?